File: py-conf/binaryHeap.py
class BinaryHeap:
    arr = []

    def __init__(self):
        self.arr = []

    def Add(self, other):
        self.arr.append(other)
        index = len(self.arr) - 1
        parent = int((index - 1) / 2)

        while index > 0 and self.arr[parent] < self.arr[index]:
            temp = self.arr[index]
            self.arr[index] = self.arr[parent]
            self.arr[parent] = temp

            index = parent
            parent = int((index - 1) / 2)

    def Heapify(self, other):
        leftChild, rightChild, largestChild = 0, 0, 0
        while True:
            leftChild = 2 * other + 1
            rightChild = 2 * other + 2
            largestChild = other

            if leftChild < len(self.arr) and self.arr[leftChild] > self.arr[largestChild]:
                largestChild = leftChild

            if rightChild < len(self.arr) and self.arr[rightChild] > self.arr[largestChild]:
                largestChild = rightChild

            if largestChild == other:
                break

            temp = self.arr[other]
            self.arr[other] = self.arr[largestChild]
            self.arr[largestChild] = temp
            other = largestChild

    def buildHeap(self, source):
        self.arr = source
        for i in range(int(len(self.arr) / 2), -1, -1):
            self.Heapify(i)

    def getMax(self):
        res = self.arr[0]
        self.arr[0] = self.arr[len(self.arr) - 1]
        self.arr.pop()
        self.Heapify(0)
        return res

File: cpp-conf/binary-heap.cpp
//
// Created by ПК on 29.05.2021.
//

#include <utility>
#include <vector>

class BinaryHeap {
private:
    std::vector<int> v;
    void heapify(int index);

public:
    BinaryHeap() {
        v = std::vector<int>(0);

    }
    explicit BinaryHeap(std::vector<int> value) {
        v = std::move(value);
    }
    void Add(int value);
    void BuildHeap(std::vector<int> source);
    void Print();
    int GetMax();
};


void BinaryHeap::Add(int value) {
    v.push_back(value);
    int index = v.size() - 1;
    int parent = (index - 1) / 2;

    while(index > 0 && v[parent] < v[index]) {
        int temp = v[index];
        v[index] = v[parent];
        v[parent] = temp;

        index = parent;
        parent = (index - 1)/ 2;
    }
}

void BinaryHeap::heapify(int index) {
    int leftChild = 0, rightChild = 0, largestChild = 0;
    for(;;) {
        leftChild = 2 * index + 1;
        rightChild = 2 * index + 2;
        largestChild = index;

        if(leftChild < v.size() && v[leftChild] > v[largestChild]) {
            largestChild = leftChild;
        }
        if(rightChild < v.size() && v[rightChild] > v[largestChild]) {
            largestChild = rightChild;
        }
        if(largestChild == index) {
            break;
        }
        int temp = v[index];
        v[index] = v[largestChild];
        v[largestChild] = temp;
        index = largestChild;
    }
}

void BinaryHeap::BuildHeap(std::vector<int> source) {
    v = std::move(source);
    for(int i = v.size() / 2; i >= 0; i--) {
        heapify(i);
    }
}

void BinaryHeap::Print() {
    for(auto i: v) {
        std::cout << i << " ";
    }
    std::cout << std::endl;
}

int BinaryHeap::GetMax() {
    int res = v[0];
    v[0] = v[v.size() - 1];
    v.pop_back();
    BinaryHeap::heapify(0);
    return res;
}

